### PR TITLE
feat(api): implement pyarrow memtables

### LINKED ISSUE #5782: feat: support creating memtables from pyarrow tables
### Is your feature request related to a problem?

`ibis.memtable()` doesn't accept a `pa.Table` and I wouldn't like to convert it to a pandas `DataFrame` before using it.

### Describe the solution you'd like

`ibis.memtable(pyarrow_table)` should be supported

### What version of ibis are you running?

5.0

### What backend(s) are you using, if any?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
This PR adds support to `ibis.memtable` for pyarrow Table objects. Closes #5782.

### CODE REVIEW

File: ibis/expr/operations/relations.py [hunk 1]
```diff
@@ -74,11 +75,7 @@ class SQLQueryResult(TableNode):
 class InMemoryTable(PhysicalTable):
     name = rlz.instance_of(str)
     schema = rlz.instance_of(sch.Schema)
-
-    @property
-    @abstractmethod
-    def data(self) -> util.ToFrame:
-        """Return the data of an in-memory table."""
+    data = rlz.instance_of(TableProxy)
```
Comment (kszucs) [inline ibis/expr/operations/relations.py:78]: Would it make sense to move the base TableProxy implementation right here, before the InMemoryTable class definition?

Comment (cpcloud) [reply ibis/expr/operations/relations.py:107]: Yep, done.

Comment (kszucs) [reply ibis/expr/operations/relations.py:107]: Thanks!

File: ibis/expr/api.py [hunk 2]
```diff
@@ -379,20 +391,69 @@ def memtable(
...
     if columns is not None and schema is not None:
         raise NotImplementedError(
             "passing `columns` and schema` is ambiguous; "
             "pass one or the other but not both"
         )
+    return _memtable_from_data(data, name=name, schema=schema, columns=columns)
+
+
+try:
+    import pyarrow as pa  # noqa: F811
```
Comment (kszucs) [inline ibis/expr/api.py:410]: Could we use `lazy_singledispatch` for `_memtable_from_data` instead to make the import optional?

Comment (cpcloud) [reply ibis/expr/api.py]: Done.

**Review by kszucs — APPROVED** [review]
Nice!

**Pull request merged**